### PR TITLE
Add Telegram and Slack attrubutes

### PR DESCRIPTION
**Pull request by arnav-ag**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 1]
```diff
@@ -23,13 +23,13 @@ public class AddCommand extends Command {
...
             + PREFIX_EMAIL + "EMAIL "
-            + PREFIX_ADDRESS + "ADDRESS "
+            + PREFIX_TELEGRAM + "ADDRESS "
             + "[" + PREFIX_TAG + "TAG]...\n"
             + "Example: " + COMMAND_WORD + " "
             + PREFIX_NAME + "John Doe "
             + PREFIX_PHONE + "98765432 "
             + PREFIX_EMAIL + "johnd@example.com "
-            + PREFIX_ADDRESS + "311, Clementi Ave 2, #02-25 "
+            + PREFIX_TELEGRAM + "311, Clementi Ave 2, #02-25 "
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/logic/commands/AddCommand.java:32]: `311, Clementi Ave 2, #02-25 ` is not an address, should we change it to telegram handle instead.

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 2]
```diff
@@ -128,7 +137,9 @@ public static class EditPersonDescriptor {
         private Name name;
         private Phone phone;
         private Email email;
-        private Address address;
+        private Telegram telegram;
+        private Slack slack;
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:141]: As we discussed before, we should use a hashMap to store this contact information such as `HashMap<ContactEnum, Contact>`

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 3]
```diff
@@ -176,14 +188,24 @@ public Optional<Email> getEmail() {
...
+        public Optional<Telegram> getTelegram() {
+            return Optional.ofNullable(telegram);
         }
 
+        public void setSlack(Slack slack) {
+            this.slack = slack;
+        }
+
+        public Optional<Slack> getSlack() {
+            return Optional.ofNullable(slack);
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:204]:
```
public void setContact(ContactEnum, Contact)
public Optional<Contact> getContact
```

These might make `Person` class more extensible in terms of managing contact.

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 4]
```diff
@@ -31,20 +31,16 @@ public class AddCommandParser implements Parser<AddCommand> {
...
     public AddCommand parse(String args) throws ParseException {
         ArgumentMultimap argMultimap =
-                ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ADDRESS, PREFIX_TAG);
+                ArgumentTokenizer.tokenize(args, PREFIX_NAME);
 
-        if (!arePrefixesPresent(argMultimap, PREFIX_NAME, PREFIX_ADDRESS, PREFIX_PHONE, PREFIX_EMAIL)
+        if (!arePrefixesPresent(argMultimap, PREFIX_NAME)
                 || !argMultimap.getPreamble().isEmpty()) {
             throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE));
         }
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java:39]: Do we still need this guard as our command line usage is `add <name>`?

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 5]
```diff
@@ -66,18 +67,33 @@ public static Phone parsePhone(String phone) throws ParseException {
...
+     * @throws ParseException if the given {@code telegram} is invalid.
      */
-    public static Address parseAddress(String address) throws ParseException {
-        requireNonNull(address);
-        String trimmedAddress = address.trim();
-        if (!Address.isValidAddress(trimmedAddress)) {
-            throw new ParseException(Address.MESSAGE_CONSTRAINTS);
+    public static Telegram parseTelegram(String telegram) throws ParseException {
+        requireNonNull(telegram);
+        String trimmedTelegram= telegram.trim();
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:77]: I think you missed a space here.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 6]
```diff
@@ -66,18 +67,33 @@ public static Phone parsePhone(String phone) throws ParseException {
...
+    /**
+     * Parses a {@code String telegram} into an {@code Address}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code telegram} is invalid.
+     */
+    public static Slack parseSlack(String slack) throws ParseException {
+        requireNonNull(slack);
+        String trimmedSlack = slack.trim();
+        if (!Telegram.isValidTelegram(trimmedSlack)) {
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:93]: Here should be validating slack account.

File: src/main/java/seedu/address/model/person/Person.java [hunk 7]
```diff
@@ -17,41 +22,67 @@ public class Person {
...
+        this.slack = slack;
+        return this;
+    }
+
+    public Telegram getTelegram() {
+        return telegram;
+    }
+    public Person setTelegram(Telegram telegram) {
+        this.telegram = telegram;
+        return this;
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/model/person/Person.java:72]: As we discussed before, we should use a hashMap to store this contact information such as `HashMap<ContactEnum, Contact>`

File: src/main/java/seedu/address/model/person/contact/Slack.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    public final String value;
+
+    /**
+     * Constructs an {@code Address}.
+     *
+     * @param channelName A valid channel name.
+     */
+    public Slack(String channelName) {
+        super(CONTACT_TYPE);
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/model/person/contact/Slack.java:26]: We can make use of `CONTACT_TYPE` in our contact hashmap.

**Review by sh4nH — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/model/person/contact/Phone.java [hunk 9]
```diff
@@ -35,19 +38,19 @@ public static boolean isValidPhone(String test) {
...
-                && value.equals(((Phone) other).value)); // state check
+            || (other instanceof Phone // instanceof handles nulls
+            && getValue().equals(((Phone) other).getValue())); // state check
     }
 
     @Override
     public int hashCode() {
-        return value.hashCode();
+        return getValue().hashCode();
     }
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/model/person/contact/Phone.java:54]: We can put these Logic to `Contact` class so we don't need to repeat every time.

File: src/main/java/seedu/address/model/person/contact/Slack.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+            || (other instanceof Telegram // instanceof handles nulls
+            && getValue().equals(((Telegram) other).getValue())); // state check
+    }
+
+    @Override
+    public int hashCode() {
+        return getValue().hashCode();
+    }
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/model/person/contact/Slack.java:59]: We can put these Logic to `Contact` class so we don't need to repeat every time.

File: src/main/java/seedu/address/model/person/contact/Slack.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        return test.matches(VALIDATION_REGEX);
+    }
+
+    /**
+     * @return
+     */
+    @Override
+    public String getLink() {
+        return SLACK_LINK_PREFIX + this.getValue();
+    }
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/model/person/contact/Slack.java:42]: Abstract `Contact` class already had this logic.

File: src/main/java/seedu/address/model/person/contact/Telegram.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        return other == this // short circuit if same object
+            || (other instanceof Telegram // instanceof handles nulls
+            && getValue().equals(((Telegram) other).getValue())); // state check
+    }
+
+    @Override
+    public int hashCode() {
+        return getValue().hashCode();
+    }
+}
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/model/person/contact/Telegram.java:71]: We can put these Logic to `Contact` class so we don't need to repeat every time.

File: src/main/java/seedu/address/model/tag/Tag.java [hunk 13]
```diff
@@ -10,7 +10,7 @@
 public class Tag {
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/model/tag/Tag.java:10]: I think `Tag` should change to `Role` instead.

File: src/main/java/seedu/address/ui/PersonListPanel.java [hunk 14]
```diff
@@ -32,7 +32,7 @@ public PersonListPanel(ObservableList<Person> personList, DetailPanelSwitcher sw
         personListView.setItems(personList);
         personListView.setCellFactory(listView -> new PersonListViewCell());
 
-        EventHandler keyEventHandler = new EventHandler<KeyEvent>() {
+        EventHandler<KeyEvent> keyEventHandler = new EventHandler<>() {
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/ui/PersonListPanel.java:35]: Thanks for correcting this error.

File: src/main/java/seedu/address/ui/ContactBox.java [hunk 15]
```diff
@@ -17,18 +25,36 @@ public class ContactBox extends UiPart<Region> {
...
         contactLabel.setText(name);
+        contactLabel.setOnAction(e -> {
+            if (Desktop.isDesktopSupported()) {
+                try {
+                    Desktop.getDesktop().browse(new URI("http://www.google.com"));
+                } catch (IOException ex) {
+                    throw new InputOutputException();
+                } catch (URISyntaxException ex) {
+                    throw new UriSyntaxParseException();
+                }
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/ui/ContactBox.java:56]: We should show an error `AlertBox` here instead.
```java
Alert a = new Alert(Alert.AlertType.ERROR);
contactLabel.setOnAction(e -> {
    if (Desktop.isDesktopSupported()) {
        try {
            Desktop.getDesktop().browse(new URI(contact.getLink()));
        } catch (IOException ex) {
            a.setContentText("Sorry, browser couldn't be opened.");
            a.show();
        } catch (URISyntaxException ex) {
            a.setContentText("Sorry, url is invalid");
            a.show();
        }
    }
});
```

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 16]
```diff
@@ -49,12 +48,11 @@ public PersonCard(Person person, int displayedIndex) {
...
-        address.setText(person.getAddress().value);
-        email.setText(person.getEmail().value);
-        person.getTags().stream()
-                .sorted(Comparator.comparing(tag -> tag.tagName))
-                .forEach(tag -> tags.getChildren().add(new Label(tag.tagName)));
+        person.getPhone().ifPresent(phone -> this.phone.setText(phone.getValue()));
+        person.getEmail().ifPresent(email -> this.email.setText(email.getValue()));
+        person.getSlack().ifPresent(slack -> this.slack.setText(slack.getValue()));
+        person.getTelegram().ifPresent(telegram -> this.telegram.setText(telegram.getValue()));
+        // person.getTag().ifPresent(tag -> this.tag.setText(tag.tagName));
```
Comment (Tex-Tang) [inline src/main/java/seedu/address/ui/PersonCard.java:55]: Any reason for this comment?

**Review by Tex-Tang — CHANGES_REQUESTED** [review]

**Pull request closed without merging**